lib/zoneinfo_module.c

```diff
@@ -0,0 +1,2194 @@
+        }
+
+        if (!dstoff && idx < (num_ttinfos - 1)) {
+            comp_idx = trans_idx[i + 1];
+
+            // If the following transition is also DST and we couldn't find
+            // the DST offset by this point, we're going to have to skip it
+            // and hope this transition gets assigned later
+            if (isdsts[comp_idx]) {
+                continue;
+            }
+
+            dstoff = utcoff - utcoffs[comp_idx];
+        }
+
+        if (dstoff) {
+            dst_found++;
+            dstoffs[idx] = dstoff;
+        }
+    }
+
+    if (dst_found < dst_count) {
+        // If there are time zones we didn't find a value for, we'll end up
+        // with dstoff = 0 for something where isdst=1. This is obviously
+        // wrong — one hour will be a much better guess than 0.
+        for (size_t idx = 0; idx < num_ttinfos; ++idx) {
+            if (isdsts[idx] && !dstoffs[idx]) {
+                dstoffs[idx] = 3600;
+            }
+        }
+    }
+}
+
+#define _swap(x, y, buffer) \
+    buffer = x;             \
+    x = y;                  \
+    y = buffer;
+
+/* Calculate transitions in local time from UTC time and offsets.
+ *
+ * We want to know when each transition occurs, denominated in the number of
+ * nominal wall-time seconds between 1970-01-01T00:00:00 and the transition in
+ * *local time* (note: this is *not* equivalent to the output of
+ * datetime.timestamp, which is the total number of seconds actual elapsed
+ * since 1970-01-01T00:00:00Z in UTC).
+ *
+ * This is an ambiguous question because "local time" can be ambiguous — but it
+ * is disambiguated by the `fold` parameter, so we allocate two arrays:
+ *
+ *  trans_local[0]: The wall-time transitions for fold=0
+ *  trans_local[1]: The wall-time transitions for fold=1
+ *
+ * This returns 0 on success and a negative number of failure. The trans_local
+ * arrays must be freed if they are not NULL.
+ */
+static int
+ts_to_local(size_t *trans_idx, int64_t *trans_utc, long *utcoff,
+            int64_t *trans_local[2], size_t num_transitions)
+{
+    if (num_transitions == 0) {
+        return 0;
+    }
+
+    // Copy the UTC transitions into each array to be modified in place later
+    for (size_t i = 0; i < 2; ++i) {
+        trans_local[i] = PyMem_Malloc(num_transitions * sizeof(int64_t));
+        if (trans_local[i] == NULL) {
+            return -1;
+        }
+
+        memcpy(trans_local[i], trans_utc, num_transitions * sizeof(int64_t));
+    }
+
+    int64_t offset_0, offset_1, buff;
+    if (num_transitions > 1) {
+        offset_0 = utcoff[0];
+        offset_1 = utcoff[trans_idx[0]];
+
+        if (offset_1 > offset_0) {
+            _swap(offset_0, offset_1, buff);
+        }
+    }
+    else {
+        offset_0 = utcoff[0];
+        offset_1 = utcoff[0];
+    }
+
+    trans_local[0][0] += offset_0;
+    trans_local[1][0] += offset_1;
+
+    for (size_t i = 1; i < num_transitions; ++i) {
```

Comment:
It's just adjusting the times for each transition from epoch times (number of UTC seconds since 1970-01-01T00:00:00Z) to be number of local "calendar-seconds" since 1970-01-01T00:00:00 in the local zone.

"Local time" is ambiguous, so there are as many as two different points where the transition could happen: one is the transitions for `fold=0` and the other is the transition when `fold=1`, so for example with a timeline with transition like this:

![Diagram of a timeline with two transitions "gap" and "fold"](https://user-images.githubusercontent.com/1377457/77542900-44a73e80-6e7d-11ea-8312-350d43e1a97a.png)

The periods of "local time" have the following pattern:

![Diagram of gaps and folds that indicates that setting fold=1 shifts the transition times over by 1 "gap/fold" period](https://user-images.githubusercontent.com/1377457/77543044-74564680-6e7d-11ea-9a2a-6a8150596ab7.png)

Which means that in the `fold=0` configuration, the transition always happens at the epoch time of the transition minus *whichever offset is larger*. If you are making the transition from UTC-4 to UTC-5 at 06:00 UTC, with `fold=0`, you want the transition to happen at 2AM (6AM - 4H), and the `fold=1` it should happen at 1AM (because the ambiguous time is from 1AM to 2AM - the doubled hours). If you are going the other way, from UTC-5 to UTC-4 at 07:00 UTC, you want the `fold=0` transition to happen at 3AM (7AM - 4H), and the `fold=1` transition to happen at 2AM, because the missing hour is from 2AM to 3AM.

So that's what this does - it looks at each transition and subtracts the larger offset from the `fold=0` transition time, then subtracts the smaller offset from the list of `fold=1` transition times.

No sorting is involved.

I will try to add a comment to explain what this is doing.